source/resource_name.c

```diff
@@ -0,0 +1,122 @@
+    }
+    prev_token_cursor->len = token_cursor->ptr - prev_token_cursor->ptr;
+    return AWS_OP_SUCCESS;
+}
+
+AWS_COMMON_API
+int aws_parse_resource_name(const struct aws_byte_cursor *in, struct aws_resource_name *out) {
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(in));
+    AWS_PRECONDITION(in->ptr);
+    AWS_PRECONDITION(out);
+
+    int errc = 0;
+    const struct aws_byte_cursor cursor_split_on = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL(ARN_DELIMETER);
+    printf("ISplit cur len: %ld\n", cursor_split_on.len);
+    printf("ISplit cur val: [" PRInSTR "]\n", AWS_BYTE_CURSOR_PRI(cursor_split_on));
+
+    struct aws_byte_cursor cursor;
+
+    if (in->len < 4) {
+        return aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+    }
+    cursor.ptr = in->ptr;
+    cursor.len = 4;
+    if (!aws_byte_cursor_eq_c_str(&cursor, "arn:")) {
+        return aws_raise_error(AWS_ERROR_MALFORMED_INPUT_STRING);
+    }
+    cursor.len = in->len;
+    aws_byte_cursor_advance(&cursor, 3);
+
+    if ((errc = s_parse_resource_name_part(&cursor, &cursor_split_on, &out->partition))) {
+        printf("We are failing here!!!\n");
+        return aws_raise_error(errc);
+    }
+    if ((errc = s_parse_resource_name_part(&cursor, &cursor_split_on, &out->service))) {
+        return aws_raise_error(errc);
+    }
+    if ((errc = s_parse_resource_name_part(&cursor, &cursor_split_on, &out->region))) {
+        return aws_raise_error(errc);
+    }
+    if ((errc = s_parse_resource_name_part(&cursor, &cursor_split_on, &out->account_id))) {
+        return aws_raise_error(errc);
+    }
+
+    out->resource_id.ptr = cursor.ptr + 1;
+    out->resource_id.len = in->len - 8 -
+        (out->partition.len + out->service.len + out->region.len + out->account_id.len);
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_COMMON_API
+int aws_resource_name_to_str(struct aws_resource_name *arn, struct aws_byte_buf *buf) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(buf));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&arn->partition));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&arn->service));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&arn->region));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&arn->account_id));
+    AWS_PRECONDITION(aws_byte_cursor_is_valid(&arn->resource_id));
+
+    const size_t arn_text_length = arn->partition.len +
+            arn->region.len +
+            arn->service.len +
+            arn->account_id.len +
+            arn->resource_id.len + 8;
+
+    if (buf->capacity < arn_text_length) {
+        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+    }
+
+    struct aws_byte_cursor prefix = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("arn:");
+    const struct aws_byte_buf colon = aws_byte_buf_from_c_str(ARN_DELIMETER);
+    struct aws_byte_cursor colon_cur;
+
+    aws_byte_cursor_read(&prefix, buf, 4);
+    aws_byte_cursor_read(&arn->partition, buf, arn->partition.len);
+    colon_cur = aws_byte_cursor_from_buf(&colon);
```

Comment:
can just set this once if you append_dynamic

Reply:
I happened to encounter the non _dynamic version first. Both make sense for what might be useful. Should there be two versions of aws_byte_buf_append_resource_name() ? One for each strategy?

Reply:
I'd go with one or the other.  You could also precalculate exactly how much room you need and then do a one-time reserve before you start appending.

Reply:
Went with aws_byte_buf_reserve_relative() and did this.